aws/resource_aws_elasticsearch_domain.go

```diff
@@ -291,6 +309,21 @@ func resourceAwsElasticSearchDomainCreate(d *schema.ResourceData, meta interface
 		}
 	}
 
+	if v, ok := d.GetOk("encrypt_at_rest"); ok {
+		options := v.([]interface{})
+
+		if len(options) > 1 {
+			return fmt.Errorf("Only a single encrypt_at_rest block is expected")
```

Comment:
I think we should also validate the instance types here since, like you mentioned, not all instance types support encryption at rest. I saw a similar approach in `t2.unlimited` [PR](https://github.com/terraform-providers/terraform-provider-aws/pull/2619/files#diff-28aade55afa0cda57279a9d35d523488R1043). I thought it was 🆒  :)

Reply:
I don't think that will run at plan time though so it's just returning a different error to what you'll get from the AWS API on apply time. You already get the following error on apply (and immediately):
```
Error: Error applying plan:

1 error(s) occurred:

* aws_elasticsearch_domain.elasticsearch: 1 error(s) occurred:

* aws_elasticsearch_domain.elasticsearch: InvalidTypeException: Encryption at rest is not supported with t2.small.elasticsearch instances
	status code: 409, request id: 085a21b5-df33-11e7-babb-3f32bfcf1d9b

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```
From memory the validatefuncs in the schema can't validate different parameters to a resource so we can't put it there unfortunately to give us a plan time failure.

Reply:
I agree with @Puneeth-n here that we should move validation to the schema here.
@tomelliff you're also right that `ValidateFunc` can't help in this context, but we can use `MaxItems: 1` which should do the job. 😉

Reply:
That's nice, means I can remove some logic there. Is it worth doing the same to some of the other blocks that are using that other style? Or (if you do want it changing) would you prefer that in a separate PR?